### ISSUE TITLE
Nested Enzyme reverse adjoint (GaussAdjoint/EnzymeVJP over an MTK DAE solve) corrupts state across repeated outer calls: first correct, subsequent fail

### ISSUE THREAD

**Issue by ChrisRackauckas-Claude** [EnzymeAD/Enzyme.jl#3247]
## Summary

Repeated reverse-mode `Enzyme.autodiff` over a SciML `solve` of an **MTK DAE**, using `GaussAdjoint(EnzymeVJP())` (so the adjoint's inner VJP is itself a nested `Enzyme.autodiff` over the MTK-generated RHS), corrupts Enzyme's global state across outer calls: the **first** outer differentiation is correct, **every subsequent** one fails. Swapping only the inner VJP to `ReverseDiffVJP()` (no nested Enzyme) makes all calls correct and identical — so the corruption is specifically the nested `EnzymeVJP` adjoint, accumulating across repeated outer passes.

This is the cross-call state-corruption cousin of the (now-fixed) BLAS-gemv #3139 and the (closed) RGF-reverse #3140 — it reproduces on Enzyme **0.13.169 / Enzyme_jll 0.0.275**, which postdate both. Downstream tracking issue: SciML/SciMLSensitivity.jl#1470 (and #1469 for the analogous init-path variant).

## Reproducer

Deterministic. Env: `Enzyme 0.13.169`, `Enzyme_jll 0.0.275`, `ModelingToolkit 11.29.0` (`ModelingToolkitBase 1.48.0`), `OrdinaryDiffEqRosenbrock 2.3.1`, `OrdinaryDiffEqNonlinearSolve 2.0.0`, `SciMLSensitivity 7.112.2`, `SciMLStructures 1.10.1`, `SciMLBase 3.24.0`, Julia 1.12.6.

```julia
using ModelingToolkit
using ModelingToolkit: t_nounits as t, D_nounits as D
using OrdinaryDiffEqRosenbrock
using OrdinaryDiffEqNonlinearSolve: BrownFullBasicInit
using SciMLBase: CheckInit
using SciMLSensitivity
import SciMLStructures as SS
using SciMLSensitivity: parameter_values
import Enzyme

@parameters σ ρ β
@variables x(t) y(t) z(t) w(t) w2(t)
eqs = [
    D(D(x)) ~ σ * (y - x), D(y) ~ x * (ρ - z) - y, D(z) ~ x * y - β * z,
    w ~ x + y + z + 2β, 0 ~ x^2 + y^2 - w2^2,          # algebraic eq -> DAE
]
@mtkbuild sys = ODESystem(eqs, t)
prob = ODEProblem(sys, [D(x) => 2.0, x => 1.0, y => 0.0, z => 0.0], (0.0, 50.0),
    [σ => 28.0, ρ => 10.0, β => 8 / 3]; jac = true, guesses = [w2 => -1.0])
tunables, repack, _ = SS.canonicalize(SS.Tunable(), parameter_values(prob))

mkloss(SA) = function (tun, prob_)
    _, rp, _ = SS.canonicalize(SS.Tunable(), parameter_values(prob_))
    np = remake(prob_; u0 = prob_.u0, p = rp(tun))
    sum(solve(np, Rodas5P(); initializealg = CheckInit(), sensealg = SA,
              abstol = 1e-6, reltol = 1e-3))
end

function repeat_outer(SA; iters = 6)        # same outer autodiff, repeated in one process
    loss = mkloss(SA)
    for i in 1:iters
        dtun = zero(tunables); diprob = Enzyme.make_zero(prob)
        try
            Enzyme.autodiff(Enzyme.set_runtime_activity(Enzyme.Reverse),
                Enzyme.Const(loss), Enzyme.Active,
                Enzyme.Duplicated(copy(tunables), dtun),
                Enzyme.Duplicated(prob, diprob))
            println("  iter $i: grad=$dtun")
        catch e
            println("  iter $i: ERROR $(typeof(e))")
        end
    end
end

println("ReverseDiffVJP (no nested Enzyme) — control:")
repeat_outer(GaussAdjoint(; autojacvec = SciMLSensitivity.ReverseDiffVJP()))
println("EnzymeVJP (nested Enzyme adjoint):")
repeat_outer(GaussAdjoint(; autojacvec = SciMLSensitivity.EnzymeVJP()))
```

### Output

```
ReverseDiffVJP (no nested Enzyme) — control:
  iter 1: grad=[-16542.990592406342, 9989.120726533434, 16244.31886896211]
  iter 2: grad=[-16542.990592406342, 9989.120726533434, 16244.31886896211]
  ... (iters 3–6 identical)
EnzymeVJP (nested Enzyme adjoint):
  iter 1: grad=[-16540.140950947123, 9991.578272706383, 16235.83418111794]   # correct
  iter 2: ERROR SciMLBase.CheckInitFailureError
  iter 3: ERROR SciMLBase.CheckInitFailureError
  ... (iters 4–6 identical error)
```

`iter 1` of `EnzymeVJP` matches the `ReverseDiffVJP` control to ~5e-4, so the nested Enzyme adjoint is *correct on the first call*. The state corruption only manifests from the **second** outer call onward. (In the full multi-setup test the same corruption surfaces as `LinearAlgebra.SingularException` out of the semi-explicit DAE adjoint's linear solve; here, with `CheckInit`, the poisoned state makes the next forward DAE initialization fail instead — same root, different downstream symptom.)

## What I could NOT reduce it to (the trigger needs the `solve`/adjoint path)

The bare nested operation does **not** reproduce — these are all correct and stable across repeats on the same versions:

- Repeated inner `Enzyme.autodiff(Reverse, Duplicated(Void(f), …), Duplicated(du, λ), Duplicated(u, du_u), Duplicated(p, dp), Const(t))` where `f` is a hand-written DAE-shaped RHS, or a raw `@RuntimeGeneratedFunction`, or the **actual MTK `GeneratedFunctionWrapper` + `MTKParameters`** — correct cotangent every iteration.
- Outer reverse over that inner reverse (reverse-over-reverse) with the same `f` — also stable.

So unlike #3140 (single nested reverse over the RGF) this is **not** reproducible from the VJP call alone; it needs the `GaussAdjoint` adjoint solve driven through DiffEqBase's custom `EnzymeRules.reverse` for `solve`, repeated across outer passes. That's consistent with the analogous init-path bug in SciML/SciMLSensitivity.jl#1469, which also only reproduced through the rrule path.

## Relationship to prior issues

- **#3139** (nested reverse over const×active BLAS gemv) — *fixed* in 0.13.163; this reproduces on 0.13.169, so it is not that bug.
- **#3140** (reverse over a `Void`-wrapped MTK `GeneratedFunctionWrapper` under nested AD) — closed; that was the *per-call* wrong/zero/segfault, where the silent-zero was ruled a ForwardDiff-over-Enzyme composition issue. This issue is the **pure-Enzyme cross-call state corruption** (no ForwardDiff involved), which #3140 did not cover.
- Downstream: SciML/SciMLSensitivity.jl#1470 pins `ReverseDiffVJP` as the inner VJP specifically to avoid this; a `@test_broken` tracking the `EnzymeVJP` sweep is in SciML/SciMLSensitivity.jl#1504.

Happy to bisect or test candidate fixes.

**Comment by ChrisRackauckas-Claude** [comment]
## Root cause pinned: reverse mutates the `Initials` buffer of the `MTKParameters` (read-only violation)

Tracked down what the corruption actually is — it's the **"reverse mutates a buffer it should treat read-only"** family (cf. #3136, #2868, #3124), and the corrupted memory is now identified exactly.

Diagnostics on the reproducer (same versions: Enzyme 0.13.169 / jll 0.0.275):

- **No user-visible buffer that I first suspected is touched:** after the (correct) first adjoint, `prob.u0 Δ=0`, the `Tunable` parameter portion `Δ=0`, `initializeprob.u0 Δ=0`.
- **The `Initials` portion of the `MTKParameters` is mutated by `Δ=2.0`** (= the `D(x)=>2.0` initial condition) on the first adjoint, and stays corrupted:
  ```
  iter 1: OK    grad=[-16540.14, 9991.58, 16235.83]
    Δ after iter 1:  u0 Δ=0.0  Tunable Δ=0.0  Initials Δ=2.0
  iter 2: ERROR CheckInitFailureError       # next solve reads the dirty Initials
  ```
- The `Initials` array is part of the `const`/`Duplicated` `MTKParameters`; the user loss's `repack(tunables)` shares it across calls, so the second solve's DAE `CheckInit` sees the wrong algebraic-variable initial → `normresid = √0.2` → failure.

### It is *not* Enzyme-global state — it lives in `prob`

A freshly-built **or `deepcopy`d `prob` per call makes all iterations correct and identical**; the corruption only carries over when `prob` (hence its `MTKParameters.Initials`) is shared. So the reverse pass is writing a wrong value into that `Initials` array rather than leaving it read-only — the gradient on the first call is still correct, exactly the #3136 signature (correct gradient, dirtied buffer).

### Where the write comes from
The GaussAdjoint backward pass re-solves the forward DAE at checkpoints (`remake(sol.prob, …)`, sharing `p`); DAE initialization writes the solved algebraic-initial back into the shared `Initials` array, and Enzyme's reverse over that in-place write lands a corrupted value. `ReverseDiffVJP` as the inner VJP never triggers it (no Enzyme reverse over that write).

So the actionable item is the same as #3136: reverse mode should not write through the `Initials` storage (it's effectively read-only input here). Happy to test a candidate fix.

**Comment by ChrisRackauckas-Claude** [comment]
For reference: I put a **SciML-side workaround** in SciML/SciMLSensitivity.jl#1507 — the GaussAdjoint adjoint now `deepcopy`s the problem parameters so the in-place write lands on a throwaway copy instead of the user's `MTKParameters.Initials`. That makes repeated differentiation correct again. But it's only papering over the underlying defect here: reverse mode writes through the `Initials` array, which is read-only input. A fix on the Enzyme side (not writing through that storage) is still the real resolution.

**Comment by wsmoses** [comment]
@ChrisRackauckas-Claude / @ChrisRackauckas : My LLM believes this to be a sciml-side problem, not enzyme:

I have diagnosed the issue and verified the fix.

Root Cause
The problem is a primal-shadow aliasing bug in SciMLSensitivity.

During the setup of the adjoint problem for EnzymeVJP in 

adjoint_common.jl
, a shadow parameter buffer _shadow_p is pre-allocated:

julia
_shadow_p = if _needs_shadow
    repack(zero(tunables))
Because the repack function (generated by SciMLStructures.canonicalize) only populates the tunable parameters from the input vector, it copies the remaining non-tunable fields (like initials or discrete parameters) by reference from the original primal p to construct the new MTKParameters struct.

This means _shadow_p.initials === p.initials (they point to the exact same array in memory).

During the adjoint solve, _vecjacobian! is called. To prepare for the inner VJP, it zeroes out the shadow parameter buffer using Enzyme.remake_zero!(_cached_shadow) (where _cached_shadow is _shadow_p). Because of the aliasing, this mutatively zeroes out the primal parameters in the user's prob object (prob.p.initials becomes all zeros). This silently poisons the prob state during the first reverse pass, causing any subsequent forward passes or DAE initializations to fail.

The Fix
The shadow parameter struct _shadow_p must be a completely independent copy of p with no shared mutable arrays.

I modified 

adjoint_common.jl
 to initialize _shadow_p using Enzyme.make_zero(p) instead:

diff
         _shadow_p = if _needs_shadow
-            repack(zero(tunables))
+            Enzyme.make_zero(p)
Enzyme.make_zero(p) recursively allocates new zeroed arrays for all fields of p, ensuring that _shadow_p is a valid, disjoint shadow of p. At the end of _vecjacobian!, canonicalize(Tunable(), _shadow_p) is still able to extract the accumulated gradients for the tunables and copy them out, while the non-tunable fields are safely zeroed without mutating the primal.

After applying this change, the reproducer successfully runs all 6 iterations with identical gradients and no state corruption.

**Comment by ChrisRackauckas-Claude** [comment]
Agreed — closing as not-an-Enzyme-bug. Your diagnosis is correct and sharper than mine: it's a primal–shadow aliasing bug in SciMLSensitivity's `adjoint_common.jl`, where the EnzymeVJP shadow parameters `_shadow_p = repack(zero(tunables))` share the non-tunable arrays (`initials`) by reference with the primal `p`, so `Enzyme.remake_zero!(_shadow_p)` zeroes the user's `p.initials`. That reconciles exactly with the `Initials Δ=2.0` I measured — the element holding the `D(x)=>2.0` initial was zeroed (2.0→0.0). Enzyme was doing the right thing on the buffer it was handed; the buffer just aliased read-only input.

Fixing it SciML-side at the shadow allocation (`make_zero(p)`), which also covers all the adjoint methods at once. Thanks for the precise root-cause.
